File: 08-stackQueues/important/astroidCollisions.cpp
#include <bits/stdc++.h>
#include <cstdlib>
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

stack<int> astroidCollisions(vector<int> arr) {
    // const n = arr.size();
    stack<int> stk;
    for (int a : arr) {
        if (a > 0)
            stk.push(a);
        else {
            while (!stk.empty() && stk.top() > 0 && stk.top() < abs(a))
                stk.pop();
            if (!stk.empty() && stk.top() == abs(a))
                stk.pop();
            else if (stk.empty())
                stk.push(a);
        }
    }
    return stk;
}

int main() {
    vector<int> arr = {4, 7, 1, 1, 2, -3, -7, 17, 15, -16};
    stack<int> stk = astroidCollisions(arr);

    while (!stk.empty()) {
        cout << stk.top() << endl;
        stk.pop();
    }
    return 0;
}

File: 08-stackQueues/important/zig/astroidCollisions.zig
const std = @import("std");
const ArrayList = std.ArrayList(i32);
const Allocator = std.mem.Allocator;
const Stack = @import("Stack.zig").Stack(i32);
const Math = std.math;
var gpa = std.heap.GeneralPurposeAllocator(.{}).init;
fn AstroidCollistion(allocator: Allocator, arr: []const i32) !Stack {
    const n = arr.len;
    var stk: Stack = try Stack.init(allocator, n);
    for (arr) |ele| {
        if (ele > 0) try stk.push(ele) else {
            while (!stk.isEmpty() and stk.top().? > 0 and stk.top().? < @abs(ele)) {
                try stk.pop();
            }
            if (!stk.isEmpty() and stk.top().? == @abs(ele)) {
                try stk.pop();
            } else if (stk.isEmpty()) {
                try stk.push(ele);
            }
        }
    }
    return stk;
}
